File: AIReaderKindle/src/Domain/Books/BookKey.cpp
#include "BookKey.hpp"

#include "../../Support/Text.hpp"

#include <vector>

namespace BookKey {

namespace {

std::string normalize(const std::string& value) {
    std::vector<std::string> words;
    for (const auto& part : Text::split(Text::lower(value), ' ')) {
        for (const auto& word : Text::split(Text::replaceAll(Text::replaceAll(part, "\t", " "), "\n", " "), ' ')) {
            if (!word.empty()) words.push_back(word);
        }
    }
    return Text::join(words, " ");
}

}  // namespace

std::string make(const std::string& title, const std::string& author) {
    return normalize(title) + "|" + normalize(author);
}

}  // namespace BookKey

File: AIReader/AIReader/Domain/Books/BookKey.swift
import Foundation

/// The name a book goes by across devices, since neither the row id nor the
/// file is the same on two of them: its title and author, normalized.
enum BookKey {
    static func make(title: String, author: String?) -> String {
        "\(normalize(title))|\(normalize(author ?? ""))"
    }

    private static func normalize(_ text: String) -> String {
        text.lowercased()
            .split(whereSeparator: \.isWhitespace)
            .joined(separator: " ")
    }
}

extension Book {
    var key: String { BookKey.make(title: title, author: author) }
}
